README.md
# apiclient

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

Simple api client wrapper without throwing exceptions with statuscodes

## Install

Via Composer

``` bash
$ composer require dasbit/apiclient
```

## Usage

``` php
$api = new dasbit\apiclient(new Guzzle, 'https://example-host.com/api');
$api->authenticate('login', 'password');
$response = $api->request('/entity-list');
$code = $response['code'];
$responseBody = $response['body']; // assoc array of decoded json string
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email abylhasov@gmail.com instead of using the issue tracker.

## Credits

- [Dastan Abylkhassov][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/dasbit/apiclient.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/dasbit/apiclient/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/dasbit/apiclient.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/dasbit/apiclient.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/dasbit/apiclient.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/dasbit/apiclient
[link-travis]: https://travis-ci.org/dasbit/apiclient
[link-scrutinizer]: https://scrutinizer-ci.com/g/dasbit/apiclient/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/dasbit/apiclient
[link-downloads]: https://packagist.org/packages/dasbit/apiclient
[link-author]: https://github.com/dasbit
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `apiclient` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/ClientTest.php
<?php

declare(strict_types=1);

namespace dasbit\apiclient;

use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\Client as Guzzle;

class ClientTest extends \PHPUnit\Framework\TestCase
{
    private function getClientInstanceWithMocks(
        array $mocks = [],
        $defaultRequestOptions = [],
        $tokenPlace = []
    ): Client {
        foreach ($mocks as $mock) {
            $responses[] = new Response(
                $mock['code'],
                $mock['headers'],
                is_array($mock['body'])
                    ? json_encode($mock['body'], JSON_HEX_APOS | JSON_HEX_QUOT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_LINE_TERMINATORS | JSON_THROW_ON_ERROR)
                    : $mock['body']
            );
        }
        $guzzle = new Guzzle(['handler' => HandlerStack::create(new MockHandler($responses))]);
        return new Client($guzzle, 'http://fake-host.com/', $defaultRequestOptions, $tokenPlace);
    }

    private function arraysAreSimilar(array $a, array $b): bool
    {
        foreach ($a as $k => $v) {
            if (is_array($v) && !$this->arraysAreSimilar($v, $b[$k])) {
                return false;
            } else {
                if (!isset($b[$k]) || $v !== $b[$k]) {
                    return false;
                }
            }
        }
        return true;
    }

    public function testAuthenticateSuccess()
    {
        $token = 'dsfd79843r32d1d3dx23d32d';
        $client = $this->getClientInstanceWithMocks([[
            'code' => 200,
            'headers' => ['Content-Type' => 'application/json'],
            'body' => [
                'status' => 'OK',
                'token' => $token,
            ],
        ]]);
        self::assertTrue($client->authenticate('test', 'test'));
        self::assertSame($token, $client->getToken());
    }

    public function testAuthenticateFail()
    {
        $client = $this->getClientInstanceWithMocks([[
            'code' => 400,
            'headers' => ['Content-Type' => 'application/json'],
            'body' => [
                'status' => 'ERROR',
            ]
        ]]);
        self::assertFalse($client->authenticate('test', 'test'));
        self::assertNull($client->getToken());
    }

    public function testRequestsSuccess()
    {
        $body =
        <<<JSONBODY
{
	"status": "OK",
	"active": "1",
	"blocked": false,
	"created_at": 1587457590,
	"id": 23,
	"name": "Ivanov Ivan",
	"permissions": [
    	{
        	"id": 1,
        	"permission": "comment"
    	},
    	{
        	"id": 2,
        	"permission": "upload photo"
    	},
    	{
        	"id": 3,
        	"permission": "add event"
    	}
	]
}
JSONBODY;

        $code = 200;
        $client = $this->getClientInstanceWithMocks([
            [
                'code' => $code,
                'headers' => ['Content-Type' => 'application/json'],
                'body' => $body
            ]
        ]);
        $body = json_decode($body, true, 512, JSON_THROW_ON_ERROR);
        $response = $client->request('get', 'fake');

        self::assertTrue($this->arraysAreSimilar($body, $response['body']));
        self::assertSame($code, $response['code']);
    }

    public function testRequestFail()
    {
        $code = 400;
        $body =  [
            'status' => "ERROR"
        ];
        $client = $this->getClientInstanceWithMocks([
            [
                'code' => $code,
                'headers' => ['Content-Type' => 'application/json'],
                'body' => $body
            ]
        ]);
        $response = $client->request('get', 'fake');
        self::assertTrue($this->arraysAreSimilar($body, $response['body']));
        self::assertSame($code, $response['code']);
    }

    public function testTokenPlace()
    {
        $body = ['salam' => 'aleikum'];
        $token = 'dsfd79843r32d1d3dx23d32d';
        $defaultRequestOptions = [
            'headers' => [
                'Content-Type' => 'application/json',
                'User-Agent' => 'SimpleHttpClient/v1',
            ]
        ];
        $tokenPlace = [
            'headers' => [
                'Authorization' => 'Bearer {token}'
            ],
        ];
        $client = $this->getClientInstanceWithMocks([
            [
                'code' => 200,
                'headers' => ['Content-Type' => 'application/json'],
                'body' => [
                    'status' => 'OK',
                    'token' => $token,
                ],
            ],
            [
                'code' => 200,
                'headers' => ['Content-Type' => 'application/json'],
                'body' => [
                    'status' => 'OK',
                    'body' => $body,
                ],
            ]
        ], $defaultRequestOptions, $tokenPlace);
        self::assertTrue($client->authenticate('fake', 'fake'));
        self::assertSame($token, $client->getToken());
        $tokenPlace['headers']['Authorization'] = str_replace('{token}', $client->getToken(), $tokenPlace['headers']['Authorization']);
        $options = array_merge_recursive($defaultRequestOptions, $tokenPlace);
        $response = $client->request('post', 'fake');
        self::assertTrue($this->arraysAreSimilar($response['options'], $options));
    }

//    public function testClientErrors()
//    {
//      //todo
//    }
}


tests/ExampleTest.php
<?php

declare(strict_types=1);

namespace dasbit\apiclient;

class ExampleTest extends \PHPUnit\Framework\TestCase
{
    /**
     * Test that true does in fact equal true
     */
    public function testTrueIsTrue()
    {
        $this->assertTrue(true);
    }
}


src/Client.php
<?php

declare(strict_types=1);

namespace dasbit\apiclient;

use Exception;
use JsonException;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Client as GuzzleClient;

class Client
{
    /**
     * Instance of http client
     *
     * @var GuzzleClient $http
     */
    private GuzzleClient $http;

    /**
     * Target host for this client
     *
     * @var string $host
     */
    private string $host;

    /**
     * Access token
     *
     * @var string|null $token
     */
    private ?string $token = null;


    /**
     * @var array|string[][] $defaultRequestOptions
     */
    protected array $defaultRequestOptions = [
        'headers' => [
            'Content-Type' => 'application/json',
            'User-Agent' => 'SimpleHttpClient/v1',
        ]
    ];

    /**
     * Token place for authenticated requests
     *
     * @var array|string[][]
     */
    protected array $tokenPlace = [
        'query' => [
            'token' => '{token}'
        ],
        //or
//        'headers' => [
//            'Authorization' => 'Bearer {token}'
//        ]
    ];

    /**
     * Client constructor.
     *
     * @param GuzzleClient $http
     * @param string $host
     * @param array $defaultRequestOptions
     * @param array $tokenPlace
     */
    public function __construct(
        GuzzleClient $http,
        string $host,
        array $defaultRequestOptions = [],
        array $tokenPlace = []
    ) {
        $this->http = $http;
        $this->host = $host;
        if (!empty($defaultRequestOptions)) {
            $this->defaultRequestOptions = $defaultRequestOptions;
        }
        if (!empty($tokenPlace)) {
            $this->tokenPlace = $tokenPlace;
        }
    }

    /**
     * Sets token
     *
     * @param string|null $token
     */
    public function setToken(?string $token): void
    {
        $this->token = $token;
    }

    /**
     * Returns token
     *
     * @return string|null
     */
    public function getToken(): ?string
    {
        return $this->token;
    }

    /**
     * Authenticates and store token
     *
     * @param $login
     * @param $password
     * @return bool
     * @throws ClientException
     */
    public function authenticate($login, $password): bool
    {
        //todo realize more flexible
        $res = $this->request('get', 'auth', [
            'query' => compact('login', 'password')
        ]);

        if ($res['code'] === 200) {
            $this->setToken($res['body']['token'] ?? null);
            return true;
        }
        return false;
    }

    /**
     * Request
     *
     * @param $method
     * @param $path
     * @param array $options
     * @return array|null
     * @throws ClientException
     */
    public function request($method, $path, array $options = []): ?array
    {
        $options = array_merge_recursive($this->defaultRequestOptions, $options);
        try {
            if ($this->token !== null) {
                $where = array_key_first($this->tokenPlace);
                $key = array_key_first($this->tokenPlace[$where]);
                $value = $this->tokenPlace[$where][$key];
                $options[$where][$key] = str_replace('{token}', $this->token, $value);
            }
            $response = $this->http->request($method, $this->host . $path, $options);
            $code = (int) $response->getStatusCode();
            $headers = $response->getHeaders();
            $body = $this->handleResponseBody((string) $response->getBody(), $headers);
            $error = null;
        } catch (RequestException $e) {
            $headers = null;
            $body = null;
            if ($e->getResponse() !== null) {
                $headers = $e->getResponse()->getHeaders();
                $body = $this->handleResponseBody((string) $e->getResponse()->getBody(), $headers);
            }
            $code = (int) $e->getCode();
            $error = $e->getMessage();
        } catch (Exception $e) {
            throw new ClientException($e->getMessage(), 1, $e);
        }
        return compact('code', 'body', 'headers', 'error', 'options');
    }

    /**
     * Async requests
     *
     * @param array $requestData
     * @return string[]
     */
    public function asyncRequests(array $requestData): array
    {
        return ['coming_soon'];
    }

    /**
     * Handles response body
     *
     * @param string $body
     * @param array $headers
     * @return mixed
     * @throws JsonException
     */
    protected function handleResponseBody(string $body, $headers = [])
    {
        if (
            isset($headers['Content-Type'])
            && in_array('application/json', $headers['Content-Type'], true)
        ) {
            return json_decode($body, true, 512, JSON_THROW_ON_ERROR);
        }
        return $body;
    }
}
